uaclient/files.py

```diff
@@ -0,0 +1,296 @@
+
+    def create_file(self):
+        file_dir = os.path.dirname(self.filepath)
+        if not os.path.exists(file_dir):
+            os.makedirs(file_dir)
+            if self._is_private:
+                os.chmod(file_dir, 0o700)
+
+
+class PublicMachineTokenFile(UAFile):
+    _file_name = defaults.PUBLIC_MACHINE_TOKEN_FILE
+    _is_private = False
+
+    def __init__(
+        self,
+        root_directory: Optional[str] = None,
+    ):
+        if root_directory:
+            self._root_directory = root_directory
+        file_name = defaults.PUBLIC_MACHINE_TOKEN_FILE
+        is_private = False
+        super().__init__(self._root_directory, file_name, is_private)
+
+    def write(self, content):
+        """
+        Writes to the public file of the machine token
+        It checks for new keys and filter the data
+        """
+        try:
+            if isinstance(content, str):
+                content = json.loads(content)
+        except Exception:
+            return None
+        self.filter_tokens(schemas.contract_schema, content)
+        content = json.dumps(content, cls=util.DatetimeAwareJSONEncoder)
+        content = content.replace('"', "'")
+        content = util.redact_sensitive_logs(content)
+        content = content.replace("'", '"')
+        super().write(content)
+
+    def filter_tokens(self, schema, data):
+        keys_to_del = []
+        for k, v in data.items():
+            if k not in schema:
+                keys_to_del.append(k)
+                LOG.debug(
+                    "{key} not present in {schema}".format(
+                        key=k, schema=schema
+                    )
+                )
+            else:
+                if isinstance(v, dict):
+                    self.filter_tokens(schema[k], v)
+                elif isinstance(v, list):
+                    for i_v in v:
+                        if isinstance(i_v, dict):
+                            self.filter_tokens(schema[k][0], i_v)
+        for k in keys_to_del:
+            LOG.debug("Deleting {k} key from the data".format(k=k))
+            if isinstance(data, Dict):
+                del data[k]
+            else:
+                if k in data:
+                    data.remove(k)
+
+
+class MachineTokenFile:
+    _machine_token = None
+    _entitlements = None
+    _contract_expiry_datetime = None
+
+    def __init__(
+        self, root_directory: Optional[str] = None, root_mode: bool = True
+    ):
+        self._rd = None
+        self._fn = None
+        if root_mode:  # private file
+            if not root_directory:
+                root_directory = defaults.DEFAULT_DATA_DIR + "/private"
+            file_name = defaults.PRIVATE_MACHINE_TOKEN_FILE
+            self._rd = root_directory
+            self._fn = file_name
+            self.file_handler = UAFile(root_directory, self._fn)
+        else:  # public file
+            self.file_handler = PublicMachineTokenFile(root_directory)
```

Comment:
If we move the `filter_tokens` logic into this class, then the public machine token file can just be a normal `UAFile` instance

Reply:
Hahah, that is true. Tested it and no issue found.